iothub/device/src/Transport/AmqpClientConnectionPool.cs

```diff
@@ -0,0 +1,155 @@
+            }
+        }
+        internal void RemoveClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+
+            if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySasSingle)
+            {
+                if (singleSasPool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    singleSasPool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from singleSasPool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityX509)
+            {
+                if (x509Pool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    x509Pool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from X509Pool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            if (Logging.IsEnabled) Logging.Exit(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+        }
+
+
+        private AmqpClientConnection GetIoTHubSasSingleClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (singleSasPool.ContainsKey(deviceClientEndpointIdentity))
+            {
+                // Error handling!!!
+                return null;
+            }
+            else
+            {
+                AmqpClientConnectionFactory amqpClientConnectionFactory = new AmqpClientConnectionFactory();
+
+                AmqpClientConnection amqpClientConnection = amqpClientConnectionFactory.Create(deviceClientEndpointIdentity, RemoveClientConnection);
+                if (singleSasPool.TryAdd(deviceClientEndpointIdentity, amqpClientConnection))
+                {
+                    return amqpClientConnection;
+                }
+                else
+                {
+                    // Error handling!!!
+                    return null;
+                }
+            }
+        }
+
+        private AmqpClientConnection GetIoTHubX509ClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (x509Pool.ContainsKey(deviceClientEndpointIdentity))
+            {
+                // Error handling!!!
+                return null;
```

Comment:
Maybe return existing connection instead?

Reply:
Yes, but this code is already obsolete. There will be no pool for single connections.

---
In reply to: [251666209](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251666209) [](ancestors = 251666209)